src/command.ts

```diff
@@ -0,0 +1,261 @@
+import { first, last, get, size, head, keys, values } from "lodash";
+import { bold } from "cli-color";
+
+import { FirebaseError } from "./error";
+import { getInheritedOption } from "./utils";
+import { load } from "./rc";
+import { load as _load } from "./config";
+import * as configstore from "./configstore";
+import detectProjectRoot = require("./detectProjectRoot");
+import logger = require("./logger");
+import track = require("./track");
+import { CommanderStatic } from "commander";
+
+type ActionFunction = (...args: any[]) => Promise<any> | any;
+
+interface BeforeFunction {
+  fn: ActionFunction;
+  args: any[];
+}
+
+/**
+ * Command is a wrapper around commander to simplify our use of promise-based
+ * actions and pre-action hooks.
+ */
+export default class Command {
+  private name = "";
+  private descriptionText = "";
+  private options: any[][];
+  private actionFn: ActionFunction = (): void => {};
+  private befores: BeforeFunction[];
+  private helpText = "";
+  private client: any;
+
+  /**
+   * @param cmd the command to create.
+   */
+  constructor(private cmd: string) {
+    this.name = first(cmd.split(" ")) || "";
+    this.options = [];
```

Comment:
Should we define these defaults inline with the field declaration?

Reply:
done - thanks